### PR TITLE
Introduce `TreeNode::exists()` API, avoid copying expressions

### PR DESCRIPTION
**Pull request by peter-toth**
## Which issue does this PR close?

Part of https://github.com/apache/arrow-datafusion/issues/8913.

## Rationale for this change

This PR adds `TreeNode::exists()` API as a common pattern to check the presence of a matching node for a predicate in a tree.
This PR also solves the issue found by @alamb here: https://github.com/apache/arrow-datafusion/pull/9913#discussion_r1554563916. Since https://github.com/apache/arrow-datafusion/pull/9913 there is no need to call `LogicalPlan::expressions()` (that copies the plan's expressions) at certain places, but iterating over the expressions with `LogicalPlan::apply_expressions()` is enough.

## What changes are included in this PR?

- Adds the `TreeNode::exists()` API.
- Replaces `contains_outer_reference` helper with a better solution.

## Are these changes tested?

Yes, with existing UTs.

## Are there any user-facing changes?

No.

### CODE REVIEW

**Comment by peter-toth** [comment]
@alamb, I think there are other places where `LogicalPlan::expressions()` can be replaced with the faster `LogicalPlan::apply_expressions()` but the change is non-trivial, so let's start with this smaller PR first.

File: datafusion/common/src/tree_node.rs [hunk 1]
```diff
@@ -292,6 +292,20 @@ pub trait TreeNode: Sized {
         )
     }
 
+    fn exists<F: FnMut(&Self) -> bool>(&self, mut f: F) -> bool {
```
Comment (alamb) [inline datafusion/common/src/tree_node.rs:295]: Can we please add some documentation to this method?

Perhaps something like

```suggestion
    /// Returns true if `f` returns true for node in the tree.
    ///
    /// Stops recursion as soon as a matching node is found
    fn exists<F: FnMut(&Self) -> bool>(&self, mut f: F) -> bool {
```

File: datafusion/optimizer/src/analyzer/subquery.rs [hunk 2]
```diff
@@ -233,13 +233,6 @@ fn check_inner_plan(
     }
 }
 
-fn contains_outer_reference(inner_plan: &LogicalPlan) -> bool {
-    inner_plan
-        .expressions()
```
Comment (alamb) [inline datafusion/optimizer/src/analyzer/subquery.rs:238]: ❤️  for removing this copy 🚀

File: datafusion/expr/src/logical_plan/plan.rs [hunk 3]
```diff
@@ -1321,6 +1321,20 @@ impl LogicalPlan {
             | LogicalPlan::Extension(_) => None,
         }
     }
+
+    pub fn contains_outer_reference(&self) -> bool {
```
Comment (alamb) [inline datafusion/expr/src/logical_plan/plan.rs:1325]: If we are going to make this a public functions, I think it should have comments. For example:

```suggestion
    /// If this node's expressions contains any references to an outer subquery
    pub fn contains_outer_reference(&self) -> bool {
```

File: datafusion/optimizer/src/decorrelate.rs [hunk 4]
```diff
@@ -91,7 +91,7 @@ impl TreeNodeRewriter for PullUpCorrelatedExpr {
                     _ => Ok(Transformed::no(plan)),
                 }
             }
-            _ if plan.expressions().iter().any(|expr| expr.contains_outer()) => {
+            _ if plan.contains_outer_reference() => {
```
Comment (alamb) [inline datafusion/optimizer/src/decorrelate.rs:94]: And another copy gone!

**Review by alamb — APPROVED** [review]
Thank you @peter-toth  -- this looks great! (and thank you for keeping the PRs small).

I have some comment suggestions, but I also think we could add them as a follow on PR too

**Comment by peter-toth** [comment]
> I took the liberty of adding the doc comments directly to this PR. Thanks again @peter-toth 

Thanks @alamb for the doc comments and the review! Actually, I wanted to ask you to do that as I'm travelling this week and have limited access to github.

**Comment by alamb** [comment]
Thank *you* @peter-toth  for all your help

**Pull request merged**